### PR TITLE
Create project structure and cleanup build warnings

### PR DESCRIPTION
**Pull request by rlupton20**
This adds a Makefile which gives incremental builds and dependency
tracking.

Adds a reproducible linux build environment via floki.

Also increases the warning levels for the C compiler, and fixes up
existing warnings.

### CODE REVIEW

File: chapter-1/bins/chap1.c [hunk 1]
```diff
@@ -1,6 +1,7 @@
 #include <stdio.h>
 #include <stdlib.h>
-#include "parse.h"
+
+#include <parse.h>
```
Comment (curlywurlycraig) [inline chapter-1/bins/chap1.c:4]: What's the diff in this context between using quotes or angle brackets?

File: chapter-1/bins/chap1.c [hunk 2]
```diff
@@ -9,6 +10,8 @@ char* formulaInput;
 int main(int argc, char** argv) {
   // Read filename from args
   char* filename = argv[1];
+  (void)filename;
+  (void)argc;
```
Comment (curlywurlycraig) [inline chapter-1/bins/chap1.c:14]: What's going on here?

**Review by curlywurlycraig — APPROVED** [review]

File: chapter-1/bins/chap1.c (hunk 2 above)
Comment (rlupton20) [reply chapter-1/bins/chap1.c:14]: I turned on `-Wall -Wextra -pendantic` to give extra compiler warnings, and `-Werror` to turn warnings into errors. Unused args are warned against, and for better or worse this is one way to circumvent that warning when variables/arguments need to be present but aren't used.

I'd prefer to figure out how to turn that particular warning off (for procedure arguments) but didn't figure out how yet. We could do something like

```
#define DECLARE_UNUSED(var) (void)(var)
```

to make the intent clearer if you prefer.

File: chapter-1/bins/chap1.c (hunk 1 above)
Comment (rlupton20) [reply chapter-1/bins/chap1.c:4]: In this context, nothing. I usually treat it as a sign of intent, where the formal interface is in `.h` files in the `include/` directory, and then included with `<>`, and local implementation oriented headers can be put in `src/` with includes using quotes.

We might want to put a directory in to make that clear, e.g.

```cpp
#include <lisp-wasm/parse.h>
```

Ref: https://gcc.gnu.org/onlinedocs/cpp/Include-Syntax.html

File: chapter-1/bins/chap1.c (hunk 2 above)
Comment (curlywurlycraig) [reply chapter-1/bins/chap1.c:14]: Not too fussed, just asking out of curiosity 😄

**Pull request merged**